### PR TITLE
Add search-related Liquid tags

### PR DESCRIPTION
**Pull request by mbland**
You can see these in action in the [search-plugin branch of 18F/hub](https://github.com/18F/hub/compare/search-plugin).

Next step would be to copy the JavaScript files to the generated site; then I think it's more or less complete.

cc: @afeld

### CODE REVIEW

**Comment by mbland** [comment]
@afeld I think this is ready for a proper look now.

**Comment by mbland** [comment]
cc: @adelevie, since Aidan's mostly out this week.

Alan, notice that #2 depends on this going in first.

**Comment by adelevie** [comment]
Blocked by https://github.com/18F/jekyll_pages_api_search/issues/3, @mbland.

File: README.md [hunk 1]
```diff
@@ -1,84 +1,100 @@
-## test_temp_file_helper Gem
+## jekyll_pages_api_search Plugin
```
Comment (afeld) [inline README.md:1]: I would use separated words here. #seo

File: README.md [hunk 2]
```diff
@@ -1,84 +1,100 @@
...
-    file_path = @temp_file_helper.mkfile(File.join('path', 'to', 'file'))
-    ...
-  end
-end
-```
+- `{% jekyll_pages_api_search_interface %}`: inserts the HTML for the search
+  box and search results
+- `{% jekyll_pages_api_search_load %}`: inserts the `<script>` tags to load
+  the search code asynchronously; the search code will then load
+  `search-index.json` asynchronously
```
Comment (afeld) [inline README.md:55]: Maybe worth an example?

File: README.md [hunk 3]
```diff
@@ -1,84 +1,100 @@
...
+anything that causes the interaction between these components to fail.
+
+Alternately, you can inspect the code of this Gem (all paths relative to
+`lib/jekyll_pages_api_search/`):
+
+- `{% jekyll_pages_api_search_interface %}`: includes `search.html`
+- `{% jekyll_pages_api_search_load %}`: generated by the `LoadSearchTag` class
+  from `tags.rb`
+- `@import "jekyll_pages_api_search";`: includes
+  `sass/jekyll_pages_api_search.scss`
```
Comment (afeld) [inline README.md:74]: You might want to link to the files on GitHub.

File: README.md [hunk 4]
```diff
@@ -1,84 +1,100 @@
...
+
+You can also add `@import "jekyll_pages_api_search";` to one of your [Sass
+assets](http://jekyllrb.com/docs/assets/) to use the default interface style.
+
+### Customization
+
+If you prefer to craft your own versions of these tags and styles, you can
+capture the output of these tags and the Sass `@import` statement, then create
+new tags or included files based on this output, careful not to change
+anything that causes the interaction between these components to fail.
```
Comment (afeld) [inline README.md:65]: Not sure that this is enough information for someone to go on...I might consider just leaving it out until we have a better idea of what customizations people would want.

File: README.md [hunk 5]
```diff
@@ -1,84 +1,100 @@
...
 
-1. the `tmp_dir` argument to `TempFileHelper.new`
-2. the `TEST_TMPDIR` environment variable
-3. `Dir.mktmpdir`
+### Developing
 
-The path to the temporary directory itself can be accessed via
-`TempFileHelper.tmpdir`.
+Building the gem requires [Node.js](https://nodejs.org/) and several Node
+packages. The `Rakefile` will prompt you to install Node.js and any packages
```
Comment (afeld) [inline README.md:92]: I might say "Running `rake` will prompt...".

File: Rakefile [hunk 6]
```diff
@@ -8,3 +8,56 @@ end
...
+  which = '/usr/bin/which'
+  opts = {[:out,:err]=>'/dev/null'}
+
+  unless system(which, 'which', opts)
+    puts [
+      "Cannot automatically check for Node.js and NPM packages on this system.",
+      "If Node.js is not installed, visit https://nodejs.org/.",
+      "If any of the following packages are not yet installed, please install",
+      "them by executing `npm install -g PACKAGE`, where `PACKAGE` is one of",
+      "the names below:"].join("\n")
```
Comment (afeld) [inline Rakefile:27]: You people and your hard line wrapping :smirk:

Comment (afeld) [inline Rakefile:27]: FYI, heredocs may be of interest.

http://en.wikibooks.org/wiki/Ruby_Programming/Here_documents

File: lib/jekyll_pages_api_search/js_copier.rb [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+require 'jekyll/static_file'
+
+module JekyllPagesApiSearch
+  class JavascriptCopier
+    def self.copy_to_site(site)
+      assets_path = File.join('assets', 'js')
+      source = File.realpath(File.join(File.dirname(__FILE__), '..', '..'))
+      begin_path = source.size + File::SEPARATOR.size
+      Dir[File.join(source, assets_path, '**', '*')].each do |f|
+        next unless File.file? f and f.end_with? '.min.js'
```
Comment (afeld) [inline lib/jekyll_pages_api_search/js_copier.rb:12]: How about

``` ruby
Dir[File.join(source, assets_path, '**', '*.min.js')].each do |f|
```

File: lib/jekyll_pages_api_search/tags.rb [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+    def render(context)
+      return @code if @code
+      baseurl = context.registers[:site].config['baseurl']
+      @code = ("<script>SEARCH_BASEURL = '#{baseurl}';</script>\n" +
+       "<script async data-main=\"#{baseurl}/assets/js/main.min\" " +
+       "src=\"#{baseurl}/assets/js/vendor/requirejs/require.min.js\">" +
+       "</script>")
+    end
+  end
```
Comment (afeld) [inline lib/jekyll_pages_api_search/tags.rb:32]: Minor: I would put each distinct class in its own file so they're easy to find.

**Comment by afeld** [comment]
Holy hell that was a big PR :sweat_smile: Nothing's a blocker, so merging now so that we can have smaller, more focused PRs going forward.

Did you mean to commit the minified (non-3rd party) JS files? Would be nice to have those compiled on the first build and be done with it, so that we don't need to remember to update them.

**Pull request merged**